ir/ir.cpp

```diff
@@ -203,13 +203,7 @@ const Type_Method* Type_Table::getApplyMethodType() const
 
 void Block::setValue(const Node* node, const CompileTimeValue* value) {
     CHECK_NULL(node);
-    auto it = constantValue.find(node);
-    if (it != constantValue.end())
-        BUG_CHECK(value == constantValue[node],
-                      "%1% already set in %2% to %3%, not %4%",
-                  node, this, value, constantValue[node]);
-    else
-        constantValue[node] = value;
+    constantValue[node] = value;        
```

Comment:
how about using equals() instead?

Reply:
Do you mean equiv? I tried, but IR::CompileTimeValue is a .def "interface" without any member functions, so you cannot check for equivalence as far as I know. I do not see an easy fix for this, seems like it is a deeper issue with how the IR is structured.

FYI, this is the error I get when trying equals:
```
third_party/p4lang_p4c/ir/ir.cpp:208:26: error: no member named 'equals' in 'IR::CompileTimeValue'
        BUG_CHECK(value->equals(*constantValue[node]),
                  ~~~~~  ^
```

Reply:
This is the code generated for CompileTimeValue:
```
/// a value that can be evaluated at compile-time
class CompileTimeValue : public virtual INode {
 public:
#line 21 "/home/mbudiu/git/p4c/ir/base.def"
    bool equiv(IR::Node const & a_) const override;
#line 24 "/home/mbudiu/git/p4c/build/ir/ir-generated.h"
};
```
So it does have an equiv.

Reply:
All interfaces really inherit from INode, so they have the INode operations.

Reply:
This part of the IR I think is actually not that useful: the Block hierarchy and CompileTimeConstants looked like a good idea, but in fact are probably not. But they are part of the P4 spec, so we cannot remove them... and several passes depend on them. They were designed to make backend writing easier, but it looks like backends don't use them much.

Reply:
Thanks for clarifying. Not sure what's going on there, but on my machine I get
```
class CompileTimeValue : public virtual INode {
};
```

Reply:
That is weird. 
But you could also just call `getNode()->equiv(*other->getNode())`.

Reply:
Or define this:
```
/// a value that can be evaluated at compile-time
interface CompileTimeValue {
    bool eq(const CompileTimeValue* a) const { return getNode()->equiv(*a->getNode()); }
}
```

Unfortunately it looks like you cannot just override equiv.

Reply:
equiv does not guarantee semantic equality, but it should be better than pointer equality. It may in fact work just fine for CompileTimeValues, since their value should not depend on the context.

Reply:
The `getNode()` trick works, thanks!
Do you have a preference between doing this inline vs defining `eq` in the IR?

Reply:
Not really. Hopefully it will give the correct result, though.

Reply:
Found a nice solution.

Regarding correctness, it at least fixes the issue that I ran into previously (sorry that I cannot provide an example), so hopefully this is strictly better than pointer equality?